share/eds/ods.go

```diff
@@ -0,0 +1,104 @@
+package eds
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+
+	cbor "github.com/ipfs/go-ipld-cbor"
+	"github.com/ipld/go-car"
+	"github.com/ipld/go-car/util"
+)
+
+var errNilReader = errors.New("ods-reader: can't create ODSReader over nil reader")
+
+// bufferedODSReader will read odsSquareSize amount of leaves from reader into the buffer.
+// It exposes the buffer to be read by io.Reader interface implementation
+type bufferedODSReader struct {
+	carReader *bufio.Reader
+	// current is the amount of CARv1 encoded leaves that have been read from reader. When current
+	// reaches odsSquareSize, bufferedODSReader will prevent further reads by returning io.EOF
+	current, odsSquareSize int
+	buf                    *bytes.Buffer
+}
+
+// ODSReader reads CARv1 encoded data from io.ReadCloser and limits the reader to the CAR header
+// and first quadrant (ODS)
+func ODSReader(carReader io.ReadCloser) (io.Reader, error) {
+	if carReader == nil {
+		return nil, errNilReader
+	}
+
+	odsR := &bufferedODSReader{
+		carReader: bufio.NewReaderSize(carReader, 4),
+		buf:       new(bytes.Buffer),
+	}
+
+	// first LdRead reads the full CAR header to determine amount of shares in the ODS
+	data, err := util.LdRead(odsR.carReader)
+	if err != nil {
+		return nil, fmt.Errorf("reading header: %v", err)
+	}
+
+	var header car.CarHeader
+	err = cbor.DecodeInto(data, &header)
+	if err != nil {
+		return nil, fmt.Errorf("invalid header: %w", err)
+	}
+
+	// car header contains both row roots and col roots which is why
+	// we divide by 4 to get the ODSWidth
+	odsWidth := len(header.Roots) / 4
```

Comment:
can we make a constant for `4` ?

Reply:
I feel like there is no need for constant if the value is physical property of square. The constant will be basically 2*2 = 4. I think there is need only for constants that represent some value (length of header) or measure (time.Minute) and this one is not the case.